File: frontend/src/lib/api/sse.ts
/**
 * SSE 流式客户端 — 解析 Chat API 的 Server-Sent Events
 */

export interface SSECallbacks {
  onToken: (content: string) => void;
  onToolStart: (tool: string, input: string) => void;
  onToolEnd: (tool: string, output: string) => void;
  onDone: () => void;
  onError: (error: Error) => void;
}

export function streamChatMessage(
  sessionId: string,
  message: string,
  callbacks: SSECallbacks
): AbortController {
  const controller = new AbortController();

  const token = localStorage.getItem('access_token');

  fetch(`/api/v1/chat/sessions/${sessionId}/messages`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'text/event-stream',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify({ message }),
    signal: controller.signal,
  })
    .then(async (response) => {
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const reader = response.body!.getReader();
      const decoder = new TextDecoder();
      let buffer = '';

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';

        let currentEvent = '';
        for (const line of lines) {
          if (line.startsWith('event: ')) {
            currentEvent = line.slice(7);
          } else if (line.startsWith('data: ')) {
            const data = line.slice(6);
            try {
              const parsed = JSON.parse(data);
              switch (currentEvent) {
                case 'token':
                  callbacks.onToken(parsed.content);
                  break;
                case 'tool_start':
                  callbacks.onToolStart(parsed.tool, parsed.input);
                  break;
                case 'tool_end':
                  callbacks.onToolEnd(parsed.tool, parsed.output);
                  break;
                case 'done':
                  callbacks.onDone();
                  break;
                case 'error':
                  callbacks.onError(new Error(parsed.message || 'Unknown error'));
                  break;
              }
            } catch {
              // Ignore malformed JSON
            }
          }
        }
      }
    })
    .catch((err) => {
      if (err.name !== 'AbortError') {
        callbacks.onError(err);
      }
    });

  return controller;
}

File: backend/test/api/routes/_sse.py
"""解析 /plan/generate SSE 事件流的测试辅助函数。"""

import json
from typing import Any


def parse_sse_event(text: str, event: str) -> dict[str, Any]:
    """从 SSE 事件流文本中提取指定事件的 data 字典；找不到则抛 AssertionError。"""
    for block in text.split("\n\n"):
        name = ""
        data_lines: list[str] = []
        for line in block.split("\n"):
            if line.startswith("event:"):
                name = line[len("event:"):].strip()
            elif line.startswith("data:"):
                data_lines.append(line[len("data:"):].strip())
        if name == event and data_lines:
            return json.loads("\n".join(data_lines))
    raise AssertionError(f"SSE 流中未找到 {event} 事件")
